server/index.js

```diff
@@ -12,34 +12,36 @@ app.get('/', (req, res) => {
 });
 
 app.get('/zagat/restaurants/:id', (req, res) => {
-  const args1 = [req.params.id];
-  db.getTitle(args1, (err1, data1) => {
-    const restaurant = data1[0];
-    if (err1) {
-      res.status(404).send('error in getting the title info');
-    } else {
-      const args2 = [restaurant.category, restaurant.neighborhood, args1[0]];
-      db.getSimilar(args2, (err2, data2) => {
-        const similar = data2;
-        if (err2) {
-          res.status(404).send('error in getting similar restaurants');
-        } else {
-          res.status(200).send({ restaurant, similar });
-        }
+  const allData = {};
+  const argsTitle = [req.params.id];
+  return db.getTitle(argsTitle)
+    .then((dataTitle) => {
+      const restaurant = dataTitle[0];
+      allData.restaurant = {
+        name: restaurant.name,
+        category: restaurant.category,
+        neighborhood: restaurant.neighborhood,
+      };
+      const argsSimilar = [restaurant.category, restaurant.neighborhood, argsTitle[0]];
+      return db.getSimilar(argsSimilar);
+    })
+    .then((dataSimilar) => {
+      allData.similar = dataSimilar;
+      const results = dataSimilar.map((item) => {
+        const argsPhotos = [item.rid];
+        return Promise.resolve(db.getPhotos(argsPhotos));
       });
-    }
-  });
-});
-
-app.get('/zagat/photos/:id', (req, res) => {
-  const args = [req.params.id];
-  db.getPhotos(args, (err, data) => {
-    if (err) {
-      res.status(404).send('error in getting their photos');
-    } else {
-      res.status(200).send(data);
-    }
-  });
+      return Promise.all(results);
+    })
+    .then((dataPhotos) => {
+      allData.photos = dataPhotos;
+    })
+    .then(() => {
+      res.status(200).send(allData);
+    })
+    .catch(() => {
+      res.status(404).send('error in getting the restaurants data');
+    });
```

Comment:
This chain works! I would still keep in mind what issues may arise with a very large database and the tradeoffs between multiple queries vs join tables.

I would also take a look at the status code you are sending back if there is an error with your database. Is 404 the error you would expect if the page loaded but information was not filled in correctly?

Reply:
@ryancmoniz I have been pondering about your comment on the status I send back in case of an error. Would it make sense to use 204, since no content meeting the request has been found? I'm still fuzzy on when to use what. Thanks 🙏